File: API/src/main/java/com/ltizzi/ecommerce/model/shoporder/ShopOrderEntity.java
package com.ltizzi.ecommerce.model.shoporder;

import com.ltizzi.ecommerce.model.cart.CartEntity;
import com.ltizzi.ecommerce.model.product.ProductEntity;
import com.ltizzi.ecommerce.model.purchase.PurchaseEntity;
import com.ltizzi.ecommerce.model.user.UserEntity;
import jakarta.persistence.*;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.SQLDelete;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.annotations.Where;

import java.math.BigDecimal;
import java.sql.Timestamp;

/**
 * @author Leonardo Terlizzi
 */

@Data
@NoArgsConstructor
@Entity
@Table(name = "shop_orders")
@SQLDelete(sql = "UPDATE shop_orders SET soft_delete = true where shop_order_id=?")
@Where(clause = "soft_delete = false")

public class ShopOrderEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long shop_order_id;

    private BigDecimal total;

    @OneToOne(fetch = FetchType.LAZY)
    private ProductEntity product;

    private int cantidad;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private UserEntity user;

    private String order_state;

    @CreationTimestamp
    @Column(updatable = false)
    private Timestamp createdAt;

    @UpdateTimestamp
    private Timestamp updatedAt;

    @Column(name = "soft_delete")
    private Boolean soft_delete = Boolean.FALSE;

}

File: Server/src/entities/ShopOrderEntity.ts
import {
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  ManyToOne,
  OneToOne,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from "typeorm";
import { ShopOrder } from "../models/ShopOrder";
import { Product } from "../models/Product";
import { User } from "../models/User";
import { ProductEntity } from "./ProductEntity";
import { UserEntity } from "./UserEntity";

@Entity("shop_orders")
export class ShopOrderEntity implements ShopOrder {
  @PrimaryGeneratedColumn()
  shop_order_id!: number;

  @Column("integer", { nullable: true })
  total!: number;

  @ManyToOne(() => ProductEntity, { eager: true, nullable: false })
  @JoinColumn({ name: "product_id", referencedColumnName: "product_id" })
  product!: Product;

  @Column()
  cantidad!: number;

  @Column()
  user_id!: number;

  @OneToOne(() => UserEntity, { eager: false, nullable: false })
  @JoinColumn({ name: "user_id" })
  user!: User;

  @Column({
    type: "varchar",
    default: "PENDING",
  })
  order_state!: string;

  @CreateDateColumn({
    type: "timestamp",
    default: () => "CURRENT_TIMESTAMP",
  })
  createdAt!: Date;

  @UpdateDateColumn({
    type: "timestamp",
    default: () => "CURRENT_TIMESTAMP",
    onUpdate: "CURRENT_TIMESTAMP",
  })
  updatedAt!: Date;

  @Column({ default: false })
  soft_delete!: boolean;
}
